pkg/boot/bzimage/kver.go

```diff
@@ -0,0 +1,193 @@
+// Copyright 2021 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+//
+// SPDX-License-Identifier: BSD-3-Clause
+//
+
+package bzimage
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"strconv"
+	"strings"
+	"time"
+)
+
+/*
+values from kernel documentation and libmagic src
+
+off val
+510 0xAA55
+514 HdrS
+526	(4 bytes) != 0x0000
+526 (2 bytes, little endian) + 0x200 -> start of null-terminated version string
+*/
+
+const kverMax = 1024 //arbitrary
+
+var (
+	ErrBootSig = errors.New("missing 0x55AA boot sig")
+	ErrBadSig  = errors.New("missing kernel header sig")
+	ErrBadOff  = errors.New("null version string offset")
+	ErrParse   = errors.New("parse error")
+)
+
+//Read kernel version string. See also: (*BZImage)Kver()
+func KVer(k io.ReadSeeker) (string, error) {
+	var buf = make([]byte, kverMax)
+	_, err := k.Seek(0, io.SeekStart)
+	if err != nil {
+		return "", err
+	}
+	_, err = k.Read(buf[:530])
+	if err != nil {
+		return "", err
+	}
+	if !bytes.Equal(buf[510:512], []byte{0x55, 0xaa}) {
+		return "", ErrBootSig
+	}
+	if string(buf[514:518]) != "HdrS" {
+		return "", ErrBadSig
+	}
+	if bytes.Equal(buf[526:530], []byte{0, 0, 0, 0}) {
+		return "", ErrBadOff
+	}
+	off := int64(binary.LittleEndian.Uint16(buf[526:528])) + 0x200
+	_, err = k.Seek(off, io.SeekStart)
+	if err != nil {
+		return "", err
+	}
+	if _, err := k.Read(buf[:]); err != nil {
+		return "", err
+	}
+	return nullterm(buf), nil
+}
+
+//Read kernel version string. See also: KVer()
+func (bz *BzImage) KVer() string {
+	if bz.Header.Kveraddr == 0 {
+		return ("(unknown)")
+	}
+	start := uint64(bz.Header.Kveraddr + 0x200)
+	bclen := uint64(len(bz.BootCode))
+	hdrlen := uint64(bz.KernelOffset) - bclen
+	bcoffs := start - hdrlen
+	if bcoffs >= bclen {
+		return ("(unknown)")
+	}
+	end := bcoffs + kverMax
+	if end > bclen {
+		end = bclen
+	}
+	return nullterm(bz.BootCode[bcoffs:end])
+}
+
+//read c string from buffer
+func nullterm(buf []byte) string {
+	var i int
+	var b byte
+	for i, b = range buf {
+		if b == 0 {
+			break
+		}
+	}
+	return string(buf[:i])
+}
+
+type KInfo struct {
```

Comment:
same here, golint and go vet will likely give hints.

Reply:
Should be fixed now.